File: SIH26011/backend/app/services/spatial_engine.py
from typing import List, Tuple, Dict, Any
from shapely.geometry import Polygon, Point
import numpy as np

class SpatialEngine:
    """
    Volumetric Cadastre Spatial Engine using Shapely 2D Footprint + Vertical Bounds (Zmin, Zmax).
    """

    @staticmethod
    def create_polygon(coords: List[List[float]]) -> Polygon:
        """Create Shapely Polygon from list of [lng, lat] coordinates."""
        if len(coords) < 3:
            raise ValueError("Polygon must have at least 3 points")
        # Ensure closed polygon
        if coords[0] != coords[-1]:
            coords = coords + [coords[0]]
        return Polygon(coords)

    @staticmethod
    def calculate_area(coords: List[List[float]]) -> float:
        """Calculate approximate ground surface area in square meters."""
        poly = SpatialEngine.create_polygon(coords)
        # Approximate conversion for lat/lon in degrees to meters (~111,000m per degree)
        lat_avg = np.mean([p[1] for p in coords])
        lat_factor = 111000.0
        lon_factor = 111000.0 * np.cos(np.radians(lat_avg))
        
        # Scale coordinates to meters relative to origin
        scaled_pts = [((p[0] - coords[0][0]) * lon_factor, (p[1] - coords[0][1]) * lat_factor) for p in coords]
        scaled_poly = Polygon(scaled_pts)
        return float(scaled_poly.area)

    @staticmethod
    def calculate_volume(coords: List[List[float]], z_min: float, z_max: float) -> float:
        """Calculate volumetric capacity (m^3) given footprint and Z range."""
        area = SpatialEngine.calculate_area(coords)
        height = max(0.0, z_max - z_min)
        return float(area * height)

    @staticmethod
    def check_containment_2d(inner_coords: List[List[float]], outer_coords: List[List[float]]) -> Tuple[bool, float]:
        """
        Check if inner polygon is fully contained inside outer polygon.
        Returns (is_contained, containment_percentage).
        """
        inner = SpatialEngine.create_polygon(inner_coords)
        outer = SpatialEngine.create_polygon(outer_coords)
        
        if outer.contains(inner):
            return True, 100.0
        
        intersection = outer.intersection(inner)
        if inner.area == 0:
            return False, 0.0
        
        pct = (intersection.area / inner.area) * 100.0
        return pct >= 99.9, float(pct)

    @staticmethod
    def check_overlap_2d(poly1_coords: List[List[float]], poly2_coords: List[List[float]]) -> Tuple[bool, float]:
        """
        Check 2D horizontal footprint overlap between two polygons.
        Returns (has_overlap, overlap_area_sqm).
        """
        p1 = SpatialEngine.create_polygon(poly1_coords)
        p2 = SpatialEngine.create_polygon(poly2_coords)
        
        if not p1.intersects(p2):
            return False, 0.0
        
        intersection = p1.intersection(p2)
        if intersection.area < 1e-7:
            return False, 0.0
            
        # Convert area to sqm
        lat_avg = np.mean([pt[1] for pt in poly1_coords])
        lat_factor = 111000.0
        lon_factor = 111000.0 * np.cos(np.radians(lat_avg))
        sqm = intersection.area * lat_factor * lon_factor
        return True, float(sqm)

    @staticmethod
    def check_vertical_overlap(z_min1: float, z_max1: float, z_min2: float, z_max2: float, tolerance: float = 0.05) -> Tuple[bool, float]:
        """
        Check vertical elevation overlap between two 3D height ranges.
        Returns (has_overlap, overlap_height_m).
        """
        overlap_min = max(z_min1, z_min2)
        overlap_max = min(z_max1, z_max2)
        overlap_height = overlap_max - overlap_min
        
        if overlap_height > tolerance:
            return True, float(overlap_height)
        return False, 0.0

    @staticmethod
    def check_3d_volumetric_overlap(
        coords1: List[List[float]], z_min1: float, z_max1: float,
        coords2: List[List[float]], z_min2: float, z_max2: float
    ) -> Tuple[bool, float]:
        """
        Check full 3D volumetric overlap between two property units.
        Returns (has_overlap, overlap_volume_cum).
        """
        has_2d, overlap_sqm = SpatialEngine.check_overlap_2d(coords1, coords2)
        if not has_2d:
            return False, 0.0
            
        has_vert, overlap_m = SpatialEngine.check_vertical_overlap(z_min1, z_max1, z_min2, z_max2)
        if not has_vert:
            return False, 0.0
            
        overlap_vol = overlap_sqm * overlap_m
        return True, float(overlap_vol)

    @staticmethod
    def get_bounding_box_3d(coords: List[List[float]], z_min: float, z_max: float) -> Dict[str, float]:
        """Compute 3D axis-aligned bounding box."""
        lngs = [p[0] for p in coords]
        lats = [p[1] for p in coords]
        return {
            "min_lng": float(min(lngs)),
            "max_lng": float(max(lngs)),
            "min_lat": float(min(lats)),
            "max_lat": float(max(lats)),
            "min_z_m": float(z_min),
            "max_z_m": float(z_max)
        }

spatial_engine = SpatialEngine()

File: backend/src/services/spatialEngine.js
import { round } from "../utils.js";

const EPSILON = 1e-10;

const close = (a, b, eps = EPSILON) =>
  Math.abs(Number(a) - Number(b)) <= eps;

const closed = (coords) => {
  if (!Array.isArray(coords) || coords.length === 0) {
    return [];
  }

  const first = coords[0];
  const last = coords.at(-1);

  if (
    close(first[0], last[0]) &&
    close(first[1], last[1])
  ) {
    return coords;
  }

  return [...coords, first];
};

const cross = (a, b, c) =>
  (b[0] - a[0]) * (c[1] - a[1]) -
  (b[1] - a[1]) * (c[0] - a[0]);

const signedArea = (coords) => {
  const polygon = closed(coords);

  if (polygon.length < 4) {
    return 0;
  }

  let sum = 0;

  for (let i = 0; i < polygon.length - 1; i += 1) {
    sum +=
      polygon[i][0] * polygon[i + 1][1] -
      polygon[i + 1][0] * polygon[i][1];
  }

  return sum / 2;
};

const polygonArea = (coords) =>
  Math.abs(signedArea(coords));

const inside = (point, a, b, orientation) =>
  orientation * cross(a, b, point) >= -EPSILON;

const lineIntersection = (a, b, c, d) => {
  const x1 = a[0];
  const y1 = a[1];
  const x2 = b[0];
  const y2 = b[1];

  const x3 = c[0];
  const y3 = c[1];
  const x4 = d[0];
  const y4 = d[1];

  const denominator =
    (x1 - x2) * (y3 - y4) -
    (y1 - y2) * (x3 - x4);

  if (Math.abs(denominator) < 1e-12) {
    return b;
  }

  const px =
    ((x1 * y2 - y1 * x2) * (x3 - x4) -
      (x1 - x2) * (x3 * y4 - y3 * x4)) /
    denominator;

  const py =
    ((x1 * y2 - y1 * x2) * (y3 - y4) -
      (y1 - y2) * (x3 * y4 - y3 * x4)) /
    denominator;

  return [px, py];
};

const clipPolygon = (subject, clip) => {
  let output = closed(subject).slice(0, -1);
  const clipPolygonPoints = closed(clip).slice(0, -1);

  if (
    output.length < 3 ||
    clipPolygonPoints.length < 3
  ) {
    return [];
  }

  const orientation =
    signedArea(clipPolygonPoints) >= 0 ? 1 : -1;

  for (
    let i = 0;
    i < clipPolygonPoints.length;
    i += 1
  ) {
    const A = clipPolygonPoints[i];
    const B =
      clipPolygonPoints[
        (i + 1) % clipPolygonPoints.length
      ];

    const input = output;
    output = [];

    if (!input.length) {
      break;
    }

    let S = input.at(-1);

    for (const E of input) {
      const endInside = inside(
        E,
        A,
        B,
        orientation
      );

      const startInside = inside(
        S,
        A,
        B,
        orientation
      );

      if (endInside) {
        if (!startInside) {
          output.push(
            lineIntersection(S, E, A, B)
          );
        }

        output.push(E);
      } else if (startInside) {
        output.push(
          lineIntersection(S, E, A, B)
        );
      }

      S = E;
    }
  }

  return output;
};

const pointInPolygon = (point, polygon) => {
  const points = closed(polygon);

  if (points.length < 4) {
    return false;
  }

  let insideFlag = false;

  for (
    let i = 0, j = points.length - 1;
    i < points.length;
    i += 1
  ) {
    const xi = points[i][0];
    const yi = points[i][1];

    const xj = points[j][0];
    const yj = points[j][1];

    const intersects =
      yi > point[1] !== yj > point[1] &&
      point[0] <
        ((xj - xi) * (point[1] - yi)) /
          (yj - yi + Number.EPSILON) +
          xi;

    if (intersects) {
      insideFlag = !insideFlag;
    }

    j = i;
  }

  return insideFlag;
};

const areaSqm = (coords) => {
  const points = closed(coords);

  if (points.length < 4) {
    return 0;
  }

  const latAverage =
    points.reduce(
      (sum, point) => sum + point[1],
      0
    ) / points.length;

  const latitudeFactor = 111000;

  const longitudeFactor =
    111000 *
    Math.cos(
      (latAverage * Math.PI) / 180
    );

  const origin = points[0];

  const projected = points.map((point) => [
    (point[0] - origin[0]) *
      longitudeFactor,
    (point[1] - origin[1]) *
      latitudeFactor,
  ]);

  return round(polygonArea(projected), 4);
};

export const spatialEngine = {
  createPolygon(coords) {
    if (
      !Array.isArray(coords) ||
      coords.length < 3
    ) {
      throw new Error(
        "Polygon must have at least 3 points"
      );
    }

    return closed(coords);
  },

  calculateArea(coords) {
    return areaSqm(coords);
  },

  calculateVolume(coords, zmin, zmax) {
    const height =
      Math.max(
        0,
        Number(zmax) - Number(zmin)
      );

    return round(
      areaSqm(coords) * height,
      4
    );
  },

  checkContainment2D(inner, outer) {
    const innerPoints =
      closed(inner).slice(0, -1);

    if (
      innerPoints.length < 3 ||
      closed(outer).length < 4
    ) {
      return [false, 0];
    }

    const contained = innerPoints.every(
      (point) =>
        pointInPolygon(point, outer)
    );

    if (contained) {
      return [true, 100];
    }

    const clipped = clipPolygon(
      inner,
      outer
    );

    const innerArea = polygonArea(inner);
    const clippedArea =
      polygonArea(clipped);

    const percentage = innerArea
      ? (clippedArea / innerArea) * 100
      : 0;

    return [
      percentage >= 99.9,
      round(percentage, 2),
    ];
  },

  checkOverlap2D(poly1, poly2) {
    const intersection = clipPolygon(
      poly1,
      poly2
    );

    if (intersection.length < 3) {
      return [false, 0];
    }

    const intersectionArea =
      polygonArea(intersection);

    if (intersectionArea < 1e-12) {
      return [false, 0];
    }

    const latAverage =
      poly1.reduce(
        (sum, point) => sum + point[1],
        0
      ) / poly1.length;

    const latitudeFactor = 111000;

    const longitudeFactor =
      111000 *
      Math.cos(
        (latAverage * Math.PI) / 180
      );

    const squareMeters =
      intersectionArea *
      latitudeFactor *
      longitudeFactor;

    return [
      squareMeters > 1e-7,
      round(squareMeters, 4),
    ];
  },

  checkVerticalOverlap(
    z1min,
    z1max,
    z2min,
    z2max,
    tolerance = 0.05
  ) {
    const overlapHeight =
      Math.min(
        Number(z1max),
        Number(z2max)
      ) -
      Math.max(
        Number(z1min),
        Number(z2min)
      );

    return overlapHeight > tolerance
      ? [true, round(overlapHeight, 4)]
      : [false, 0];
  },

  check3DVolumetricOverlap(
    c1,
    z1min,
    z1max,
    c2,
    z2min,
    z2max
  ) {
    const [hasHorizontalOverlap, area] =
      this.checkOverlap2D(c1, c2);

    if (!hasHorizontalOverlap) {
      return [false, 0];
    }

    const [hasVerticalOverlap, height] =
      this.checkVerticalOverlap(
        z1min,
        z1max,
        z2min,
        z2max
      );

    if (!hasVerticalOverlap) {
      return [false, 0];
    }

    return [
      true,
      round(area * height, 4),
    ];
  },

  getBoundingBox3D(
    coords,
    zmin,
    zmax
  ) {
    if (
      !Array.isArray(coords) ||
      coords.length === 0
    ) {
      throw new Error(
        "Coordinates are required"
      );
    }

    const longitudes = coords.map(
      (point) => point[0]
    );

    const latitudes = coords.map(
      (point) => point[1]
    );

    return {
      min_lng: Math.min(...longitudes),
      max_lng: Math.max(...longitudes),
      min_lat: Math.min(...latitudes),
      max_lat: Math.max(...latitudes),
      min_z_m: Number(zmin),
      max_z_m: Number(zmax),
    };
  },

  polygonArea,
};